File: models/Gpt-3.5-turbo/cpp/code/problems/additional_problems/Special_Substrings.cpp

#include <iostream>
#include <vector>

using namespace std;

int countSpecialSubstrings(string s) {
    int n = s.length();
    vector<int> count(26), freq(n + 1);
    int left = 0, right = 0;
    int specialCount = 0;

    while (left < n) {
        while (right < n && (right == left || count[s[right] - 'a'] + 1 == count[s[right - 1] - 'a'])) {
            count[s[right] - 'a']++;
            freq[count[s[right] - 'a']]++;
            right++;
        }

        specialCount += freq[left + 1];

        int k = 2;
        while (left + k <= right && count[s[left + k - 1] - 'a'] == count[s[left + k - 2] - 'a']) {
            specialCount++;
            k++;
        }

        count[s[left] - 'a']--;
        freq[count[s[left] - 'a']]--;

        left++;
    }

    return specialCount;
}

int main() {
    string s;
    cin >> s;

    int result = countSpecialSubstrings(s);
    cout << result << endl;

    return 0;
}

File: models/Gpt-3.5-turbo/py/code/additional_problems/Special_Substrings.py

def count_special_substrings(s):
    n = len(s)
    count = 0

    # Iterate through each character as the center of the window
    for i in range(n):
        # Initialize a dictionary to keep track of character frequencies
        freq = {}

        # Expand the window to the right
        j = i
        while j < n:
            # Increment the frequency of the current character
            freq[s[j]] = freq.get(s[j], 0) + 1

            # Check if all characters in the window have the same frequency
            if len(freq.values()) == len(set(freq.values())):
                count += 1

            # Move to the next character
            j += 1

    return count

# Test the function with the given example
s = "abccabab"
print(count_special_substrings(s))
